src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs

```diff
@@ -78,22 +80,54 @@ internal class CompletionHandler : IRequestHandler<LSP.CompletionParams, LSP.Com
             var completionTrigger = await ProtocolConversions.LSPToRoslynCompletionTriggerAsync(request.Context, document, position, cancellationToken).ConfigureAwait(false);
 
             var list = await completionService.GetCompletionsAsync(document, position, completionTrigger, options: completionOptions, cancellationToken: cancellationToken).ConfigureAwait(false);
-            if (list == null)
+            if (list == null || list.Items.IsEmpty)
             {
                 return null;
             }
 
             var lspVSClientCapability = context.ClientCapabilities?.HasVisualStudioLspCapability() == true;
-
             var commitCharactersRuleCache = new Dictionary<ImmutableArray<CharacterSetModificationRule>, ImmutableArray<string>>();
 
             // Cache the completion list so we can avoid recomputation in the resolve handler
             var resultId = await _completionListCache.UpdateCacheAsync(list, cancellationToken).ConfigureAwait(false);
 
+            // Feature flag to enable the return of TextEdits instead of InsertTexts (will increase payload size).
+            // Flag is defined in VisualStudio\Core\Def\PackageRegistration.pkgdef.
+            // We also check against the CompletionOption for test purposes only.
+            Contract.ThrowIfNull(context.Solution);
+            var featureFlagService = context.Solution.Workspace.Services.GetRequiredService<IExperimentationService>();
+            var returnTextEdits = featureFlagService.IsExperimentEnabled("Roslyn.LSP.Completion") ||
+                completionOptions.GetOption(CompletionOptions.UseLSPPrototypeBehavior, document.Project.Language);
+
+            SourceText? documentText = null;
+            TextSpan? defaultSpan = null;
+            LSP.Range? defaultRange = null;
+            if (returnTextEdits)
+            {
+                // We want to compute the document's text just once.
+                documentText = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+
+                var completionChange = await completionService.GetChangeAsync(
+                    document, list.Items.First(), cancellationToken: cancellationToken).ConfigureAwait(false);
+
+                // If possible, we want to compute the item's range just once as well.
+                // Individual items can override this range later.
+                defaultSpan = completionChange.TextChange.Span;
+                defaultRange = ProtocolConversions.TextSpanToRange(defaultSpan.Value, documentText);
+            }
```

Comment:
is it really necessary to avoid doing this work in the `!returnTextEdits` case.  Seems like we could just perform it always as it woudl be fast to do this for one item.

Reply:
gotcha.  good reason :)